searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp

```diff
@@ -0,0 +1,367 @@
+        GenerationHandler::Guard read_guard;
+
+        TwoPhaseUpdateTask(Stressor &p, uint32_t d, ConstVectorRef v, ReadGuard r)
+            : parent(p), docid(d), vec(v), read_guard(r)
+        {}
+
+        void run() override {
+            auto v = vespalib::tensor::TypedCells(vec);
+            auto up = parent.index->prepare_add_document(docid, v, read_guard);
+            EXPECT_EQ(bool(up), true);
+            auto task = std::make_unique<CompleteUpdateTask>(parent, docid, vec, std::move(up));
+            auto r = parent.write_thread.execute(std::move(task));
+            if (r) {
+                fprintf(stderr, "Failed posting complete remove task!");
+                abort();
+            }
+        }
+    };
+
+    Stressor()
+        : loaded_vectors(),
+          in_progress(BitVector::create(NUM_POSSIBLE_DOCS)),
+          existing_ids(BitVector::create(NUM_POSSIBLE_DOCS)),
+          rng(),
+          vectors(),
+          gen_handler(),
+          index(),
+          multi_prepare_workers(10, 128*1024, 50),
+          write_thread(1, 128*1024, 15)
+    {
+        loaded_vectors.load();
+    }
+
+    ~Stressor() {}
+
+    void init() {
+        uint32_t m = 16;
+        index = std::make_unique<HnswIndex>(vectors, std::make_unique<FloatSqEuclideanDistance>(),
+                                            std::make_unique<InvLogLevelGenerator>(m),
+                                            HnswIndex::Config(2*m, m, 200, true));
+        vectors.set(NUM_POSSIBLE_DOCS, loaded_vectors[0]);
+    }
+    size_t get_rnd(size_t size) {
+        return rng.nextUniform() * size;
+    }
+    void add_document(uint32_t docid) {
+        size_t vec_num = get_rnd(loaded_vectors.size());
+        ConstVectorRef vec = loaded_vectors[vec_num];
+        auto guard = take_read_guard();
+        auto task = std::make_unique<TwoPhaseAddTask>(*this, docid, vec, guard);
+        auto r = multi_prepare_workers.execute(std::move(task));
+        ASSERT_EQ(bool(r), false);
+    }
+    void remove_document(uint32_t docid) {
+        auto guard = take_read_guard();
+        auto task = std::make_unique<TwoPhaseRemoveTask>(*this, docid, guard);
+        auto r = multi_prepare_workers.execute(std::move(task));
+        ASSERT_EQ(bool(r), false);
+    }
+    void update_document(uint32_t docid) {
+        size_t vec_num = get_rnd(loaded_vectors.size());
+        ConstVectorRef vec = loaded_vectors[vec_num];
+        auto guard = take_read_guard();
+        auto task = std::make_unique<TwoPhaseUpdateTask>(*this, docid, vec, guard);
+        auto r = multi_prepare_workers.execute(std::move(task));
+        ASSERT_EQ(bool(r), false);
+    }
+    void commit(uint32_t docid) {
+        index->transfer_hold_lists(gen_handler.getCurrentGeneration());
+        gen_handler.incGeneration();
+        gen_handler.updateFirstUsedGeneration();
+        index->trim_hold_lists(gen_handler.getFirstUsedGeneration());
+        in_progress->clearBit(docid);
```

Comment:
Is this racing with what happens in gen_operation(). Should we protect it with a mutex?